src/arch/userspace.rs

```diff
@@ -31,6 +31,10 @@ static HOST_CTX: Mutex<VirtContext> = Mutex::new(VirtContext::new(
     },
 ));
 
+pub fn return_userspace_ctx() -> VirtContext {
+    *HOST_CTX.lock()
+}
```

Comment:
That smells fishy, because you are not supposed to be able to return a `VirtContext` like that, and indeed after investigation it's because you made VirtContext `Copy`, which we don't want to (see comment there)

There are two options:
1. We make VirtContext `Clone` (but *not* `Copy`), and we clone it there
2. You return a `MutexGuart<VirtContext>` which is basically a wrapper around `VirtContext` but will release the lock when dropped.

I think you probably want option 1 here, so that you release the lock after the clone (otherwise any CSR write with the lock held will deadlock).